### DISCUSSION TITLE
Importing data from other apps

### DISCUSSION THREAD

**Discussion by ChloeSzilas** [iSoron/uhabits#1243] [Help & FAQ]
Is it possible to import data from other apps? I've recently switched from iOS to Android and have 2 years worth of data from my old habit app I'd like to transfer across. The app says it supports files from Tickmate, HabitBull or Rewire,  but is there a way to import a file generated from a different app / if so what are the steps to convert it to the right format?

**Comment by hiqua** [comment]
Which app is it? I guess it won't be that obvious to convert it if it's not already supported.

You could try to convert the file to one of the supported format, if the data is similar then that should be enough. If you have a sample we could have a look.

**Reply by tcramm0nd** [reply]
> You could try to convert the file to one of the supported format

I'm in a similar boat - would you be able to point me to what the expected data format for imports is? Having trouble finding it in the docs

**Comment by hiqua** [comment]
The importers are there: https://github.com/iSoron/uhabits/tree/dev/uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io

Some of the files that can be imported are there: https://github.com/iSoron/uhabits/tree/dev/uhabits-core/assets/test e.g. see the habitbull example.

The description in the settings in the `Import data` line lists the formats that are currently supported.

**Comment by fanis** [comment]
Coming back to this as I wanted to import from Trackendar and, since there's no direct importer I converted Trackendar's CSV export to something that [HabitBullCSVImporter.kt](https://github.com/iSoron/uhabits/blob/dev/uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitBullCSVImporter.kt) can handle:

I'm happy to report that all works as expected, apart from handling decimals. Every number is expected to be an int (`parseInt()` is used), whereas a very common thing to track is weight with 1 decimal.

@hiqua any thoughts on how to best handle this case?

**Reply by hiqua** [reply]
> Every number is expected to be an int (parseInt() is used)

I don't think that's quite true given this:
https://github.com/iSoron/uhabits/blob/dev/uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitBullCSVImporter.kt#L85-L89

What does your CSV look like?

**Comment by fanis** [comment]
Won't line 81 above attempt to make it into an integer first and throw an
exception, setting it to 0? I am not too familiar with Kotlin so apologies
if I'm misreading it:
https://github.com/iSoron/uhabits/blob/dev/uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitBullCSVImporter.kt#L81

```csv
HabitName,HabitDescription,HabitCategory,CalendarDate,Value,CommentText
Temp,The Temp,Signals,2023-08-11 11:39,21.2,
```
